flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -266,22 +262,18 @@ class FIRToLLVMTypeConverter : public mlir::LLVMTypeConverter {
 
   // fir.array<c ... :any>  -->  llvm<"[...[c x any]]">
   mlir::LLVM::LLVMType convertSequenceType(fir::SequenceType seq) {
-    if (!seq.hasConstantInterior())
-      llvm_unreachable("cannot lower type to LLVM IR");
     auto baseTy = unwrap(convertType(seq.getEleTy()));
-    auto shape = seq.getShape();
     auto constRows = seq.getConstantRows();
-    if (constRows) {
+    if (seq.hasConstantInterior() && constRows) {
       decltype(constRows) i = constRows;
+      auto shape = seq.getShape();
       for (auto e : shape) {
         baseTy = mlir::LLVM::LLVMType::getArrayTy(baseTy, e);
         if (--i == 0)
           break;
       }
-      if (seq.hasConstantShape())
-        return baseTy;
     }
-    return baseTy.getPointerTo();
+    return baseTy;
```

Comment:
What happens when lowering `!fir.array<?xT>` when it is not wrapped with a `!fir.ref<>`?

It should be legal to write this fir:
```mlir
func @test(%arr : !fir.array<?xf32>)
```

Reply:
It is not handled. I am not sure I follow what `!fir.array<?xf32>` on its own is representing, how could an array ssa value have unknown shape ? For which Fortran code would such fir type be needed ?